### PR TITLE
[SPARK-24107][CORE] ChunkedByteBuffer.writeFully method has not reset the limit value

### PR DESCRIPTION
**Pull request by jinhai-cloud**
JIRA Issue: https://issues.apache.org/jira/browse/SPARK-24107?jql=text%20~%20%22ChunkedByteBuffer%22

ChunkedByteBuffer.writeFully method has not reset the limit value. When 
chunks larger than bufferWriteChunkSize, such as 80 * 1024 * 1024 larger than
config.BUFFER_WRITE_CHUNK_SIZE(64 * 1024 * 1024)，only while once, will lost 16 * 1024 * 1024 byte

### CODE REVIEW

**Comment by kiszk** [comment]
Would it be possible to add a unit test?

**Comment by manbuyun** [comment]
test("writeFully() does not affect original buffer's position") {
    val chunkedByteBuffer = new ChunkedByteBuffer(Array(ByteBuffer.allocate(80*1024*1024)))
    chunkedByteBuffer.writeFully(new ByteArrayWritableChannel(chunkedByteBuffer.size.toInt))
    assert(chunkedByteBuffer.getChunks().head.position() === 0)
  }

**Comment by Ngone51** [comment]
@manbuyun you need to add the unit test into `ChunkedByteBufferSuite.scala` and push a new commit.

**Review by Ngone51 — APPROVED** [review]
Better with a  unit test.

**Comment by maropu** [comment]
Plz add `[CORE]` in the title.

File: core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala [hunk 1]
```diff
@@ -56,6 +56,12 @@ class ChunkedByteBufferSuite extends SparkFunSuite {
     assert(chunkedByteBuffer.getChunks().head.position() === 0)
   }
 
+  test("writeFully() does not affect original buffer's position") {
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala]: Hi @manbuyun .You should add a new unit test to support your own change. For example, "writeFully() can write buffer which is larger than `bufferWriteChunkSize` correctly. " And update the test code.

Comment (manbuyun) [reply core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala]: Done. Thanks

File: core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala [hunk 2]
```diff
@@ -56,6 +56,12 @@ class ChunkedByteBufferSuite extends SparkFunSuite {
     assert(chunkedByteBuffer.getChunks().head.position() === 0)
   }
 
+  test("writeFully() can write buffer which is larger than bufferWriteChunkSize correctly") {
+    val chunkedByteBuffer = new ChunkedByteBuffer(Array(ByteBuffer.allocate(80*1024*1024)))
+    chunkedByteBuffer.writeFully(new ByteArrayWritableChannel(chunkedByteBuffer.size.toInt))
+    assert(chunkedByteBuffer.getChunks().head.position() === 0)
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala]: This assert is unnecessary for this PR change. Please replace it with assert channel's length here.

File: core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala [hunk 3]
```diff
@@ -56,6 +56,12 @@ class ChunkedByteBufferSuite extends SparkFunSuite {
     assert(chunkedByteBuffer.getChunks().head.position() === 0)
   }
 
+  test("writeFully() can write buffer which is larger than bufferWriteChunkSize correctly") {
+    val chunkedByteBuffer = new ChunkedByteBuffer(Array(ByteBuffer.allocate(80*1024*1024)))
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala]: nit: space beside `*`.

File: core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala [hunk 4]
```diff
@@ -56,6 +56,12 @@ class ChunkedByteBufferSuite extends SparkFunSuite {
     assert(chunkedByteBuffer.getChunks().head.position() === 0)
   }
 
+  test("writeFully() can write buffer which is larger than bufferWriteChunkSize correctly") {
```
Comment (kiszk) [inline core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala]: nit: Would it be possible to add `SPARK-24107: ` into the start of the string? It would help us connect a UT with JIRA entry.

Comment (manbuyun) [reply core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala]: Done. Thanks

File: core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala (hunk 3 above)
Comment (manbuyun) [reply core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala]: Done. Thanks

File: core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala [hunk 5]
```diff
@@ -56,6 +56,12 @@ class ChunkedByteBufferSuite extends SparkFunSuite {
     assert(chunkedByteBuffer.getChunks().head.position() === 0)
   }
 
+  test("SPARK-24107: writeFully() write buffer which is larger than bufferWriteChunkSize") {
+    val chunkedByteBuffer = new ChunkedByteBuffer(Array(ByteBuffer.allocate(80 * 1024 * 1024)))
+    chunkedByteBuffer.writeFully(new ByteArrayWritableChannel(chunkedByteBuffer.size.toInt))
+    assert(chunkedByteBuffer.size === (80L * 1024L * 1024L))
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala]: `ByteArrayWritableChannel `'s size, not `chunkedByteBuffer`'s size.

Comment (manbuyun) [reply core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala]: My mistake, has been fixed. Thanks

**Comment by Ngone51** [comment]
cc @kiszk @maropu @cloud-fan @jiangxb1987

File: core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala [hunk 6]
```diff
@@ -63,10 +63,12 @@ private[spark] class ChunkedByteBuffer(var chunks: Array[ByteBuffer]) {
    */
   def writeFully(channel: WritableByteChannel): Unit = {
     for (bytes <- getChunks()) {
+      val limit = bytes.limit()
       while (bytes.remaining() > 0) {
```
Comment (maropu) [inline core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala]: This is not related to this pr though, `while (bytes.hasRemaining) {`?

File: core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala [hunk 7]
```diff
@@ -56,6 +56,13 @@ class ChunkedByteBufferSuite extends SparkFunSuite {
     assert(chunkedByteBuffer.getChunks().head.position() === 0)
   }
 
+  test("SPARK-24107: writeFully() write buffer which is larger than bufferWriteChunkSize") {
+    val chunkedByteBuffer = new ChunkedByteBuffer(Array(ByteBuffer.allocate(80 * 1024 * 1024)))
```
Comment (maropu) [inline core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala]: Can you configure `bufferWriteChunkSize` explicitly?

File: core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala [hunk 8]
```diff
@@ -63,10 +63,12 @@ private[spark] class ChunkedByteBuffer(var chunks: Array[ByteBuffer]) {
    */
   def writeFully(channel: WritableByteChannel): Unit = {
     for (bytes <- getChunks()) {
+      val limit = bytes.limit()
```
Comment (maropu) [inline core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala]: How about renaming `limit` to `curChunkLimit`?

File: core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala (hunk 7 above)
Comment (manbuyun) [reply core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala]: I have modified.Please check

File: core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala [hunk 9]
```diff
@@ -56,6 +56,15 @@ class ChunkedByteBufferSuite extends SparkFunSuite {
     assert(chunkedByteBuffer.getChunks().head.position() === 0)
   }
 
+  test("SPARK-24107: writeFully() write buffer which is larger than bufferWriteChunkSize") {
+    val bufferWriteChunkSize = Option(SparkEnv.get).map(_.conf.get(config.BUFFER_WRITE_CHUNK_SIZE))
+            .getOrElse(config.BUFFER_WRITE_CHUNK_SIZE.defaultValue.get).toInt
```
Comment (maropu) [inline core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala]: How about setting this value via `spark.buffer.write.chunkSize`?

Comment (manbuyun) [reply core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala]: Ok. I have added. Please check

File: core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala [hunk 10]
```diff
@@ -63,10 +63,12 @@ private[spark] class ChunkedByteBuffer(var chunks: Array[ByteBuffer]) {
    */
   def writeFully(channel: WritableByteChannel): Unit = {
     for (bytes <- getChunks()) {
-      while (bytes.remaining() > 0) {
+      val curChunkLimit = bytes.limit()
+      while (bytes.hasRemaining) {
         val ioSize = Math.min(bytes.remaining(), bufferWriteChunkSize)
         bytes.limit(bytes.position() + ioSize)
         channel.write(bytes)
+        bytes.limit(curChunkLimit)
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala]: I would rewrite this using:
```
try {
	val ioSize = Math.min(bytes.remaining(), bufferWriteChunkSize)
	bytes.limit(bytes.position() + ioSize)
	channel.write(bytes)
} finally {
	bytes.limit(curChunkLimit)
}
```
to be safe.

Comment (manbuyun) [reply core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala]: Right. When channel write throw IOException

Comment (manbuyun) [reply core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala]: I have commit this modified

File: core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala [hunk 11]
```diff
@@ -20,12 +20,12 @@ package org.apache.spark.io
 import java.nio.ByteBuffer
 
 import com.google.common.io.ByteStreams
-
-import org.apache.spark.SparkFunSuite
+import org.apache.spark.{SparkFunSuite, SharedSparkContext}
```
Comment (xuanyuanking) [inline core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala]: move SharedSparkContext before SparkFunSuite

File: core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala [hunk 12]
```diff
@@ -20,12 +20,12 @@ package org.apache.spark.io
 import java.nio.ByteBuffer
 
 import com.google.common.io.ByteStreams
```
Comment (xuanyuanking) [inline core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala]: add an empty line behind 22 to separate spark and third-party group.

**Review by xuanyuanking — COMMENTED** [review]
nit for style check

File: core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala (hunk 11 above)
Comment (manbuyun) [reply core/src/test/scala/org/apache/spark/io/ChunkedByteBufferSuite.scala]: I have fixed and commit. Thanks

**Comment by cloud-fan** [comment]
the R test is a known issue, I'm merging in to master and 2.3, thanks!

**Comment by dongjoon-hyun** [comment]
Hi, All.
I created [SPARK-24152](https://issues.apache.org/jira/browse/SPARK-24152) because we start to merge by ignoring that known unknown SparkR failure.

File: core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala [hunk 13]
```diff
@@ -63,10 +63,15 @@ private[spark] class ChunkedByteBuffer(var chunks: Array[ByteBuffer]) {
...
-        val ioSize = Math.min(bytes.remaining(), bufferWriteChunkSize)
-        bytes.limit(bytes.position() + ioSize)
-        channel.write(bytes)
+      val curChunkLimit = bytes.limit()
+      while (bytes.hasRemaining) {
+        try {
+          val ioSize = Math.min(bytes.remaining(), bufferWriteChunkSize)
+          bytes.limit(bytes.position() + ioSize)
+          channel.write(bytes)
+        } finally {
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala]: I don't think we need the `try` and `finally` here because `getChunks()` returns duplicated ByteBuffers which have their own position and limit.

File: core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala [hunk 14]
```diff
@@ -63,10 +63,15 @@ private[spark] class ChunkedByteBuffer(var chunks: Array[ByteBuffer]) {
...
     for (bytes <- getChunks()) {
-      while (bytes.remaining() > 0) {
-        val ioSize = Math.min(bytes.remaining(), bufferWriteChunkSize)
-        bytes.limit(bytes.position() + ioSize)
-        channel.write(bytes)
+      val curChunkLimit = bytes.limit()
+      while (bytes.hasRemaining) {
+        try {
+          val ioSize = Math.min(bytes.remaining(), bufferWriteChunkSize)
+          bytes.limit(bytes.position() + ioSize)
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala]: The rationale for the `limit()` isn't super-clear, but that was a problem in the original PR which introduced the bug (#18730). I'm commenting here only for cross-reference reference for folks who come across this patch in the future. I believe that the original motivation was http://www.evanjones.ca/java-bytebuffer-leak.html

File: core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala (hunk 13 above)
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala]: Do you mean this is not a real bug that can cause real workload to fail?

**Comment by JoshRosen** [comment]
No, I mean that the code here can simply follow the write call as straight
through code. We don't need to guard against exceptions here because the
duplicate of the buffer is used only by a single thread, so you can omit
the try block and just concatenate the try contents to the finally
contents. Minor bit but I wanted to comment because I initially was
confused about when errors could occur and thread safety / sharing until I
realized that the modified state does not escape this method.

File: core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala (hunk 13 above)
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala]: I think the problem is, `bytes.limit(bytes.position() + ioSize)` will change the result of `bytes.hasRemaining`, so we have to restore the limit in each loop.

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala]: I get your point. if there is an exception, there is no next loop and we don't need to restore the limit. so try finally is not needed
